Fix custom resource icons not appearing in file system

Atlinx opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

Fixes #32706 

This PR was implemented using the advice given in my original PR (#51089) along with suggestions made by @willnationsdev. It uses the existing `file_cache` inside of `editor_file_system.cpp` to cache the script attached to each resource. Additionally, this PR manually inspects each resource file for the `ext_resource` tag of type `Script` when fetching the script attached to that resource. This means we avoid unnecessary loading of subresources, etc.

Note that this implementation supports changing icons while inside the editor -- the changes will reflect when the user clicks on the filesystem again after setting a new icon.

Before
![before_fx](https://user-images.githubusercontent.com/25368491/165871483-9752b620-9c0a-46a1-ad78-aa3c0fec5502.png)

After
![image](https://user-images.githubusercontent.com/25368491/165871464-b18b8ab2-36a0-402f-aa44-a8cd1e422bba.png)

Atlinx committed:
Initial fix by adding field to FileInfo

WIP new filesystem_cache fix

Add custom resource icons to fs using filesystem cache

Fix clang format

Revert "Initial fix by adding field to FileInfo"

This reverts commit 3975b5a0a03fc1f1f9b6b97a9c28bad225415b1d.

2nd attempt using FileInfo.deps

Add willnationsdev's feedback
- Refactored code to use ClassDB for inheritance checking.
- Removed redundant writing to fi.path

Custom resource icons using dependencies approach

Revert "2nd attempt using FileInfo.deps"

This reverts commit 43bc0f4e6bcd8d4da260cba1ad46d2b4cd11f2dd.

Revert "Revert "Initial fix by adding field to FileInfo""

This reverts commit 6a810fd2cef1b1d8a7a1e96ac505c34e201acf1b.

WIP using resource UID and custom FileInfo field

Custom resource icons using dependencies approach

Add willnationsdev's feedback

- Refactored code to use ClassDB for inheritance checking.
- Removed redundant writing to fi.path

Attempt saving UID of script

This attempt failed because there's no way of fetching the UID of a script file. I assume this is because scripts aren't treated as a resource.

Revert back to script path approach

Hit a tough race condition

See editor_file_system.cpp:, line 1315, "TODO: Fix this workaround for a race condition."

Fix race condition & optimize filesystem_changed signal

Atlinx committed:
Initial fix by adding field to FileInfo

WIP new filesystem_cache fix

Add custom resource icons to fs using filesystem cache

Fix clang format

Revert "Initial fix by adding field to FileInfo"

This reverts commit 3975b5a0a03fc1f1f9b6b97a9c28bad225415b1d.

2nd attempt using FileInfo.deps

Add willnationsdev's feedback
- Refactored code to use ClassDB for inheritance checking.
- Removed redundant writing to fi.path

Custom resource icons using dependencies approach

Revert "2nd attempt using FileInfo.deps"

This reverts commit 43bc0f4e6bcd8d4da260cba1ad46d2b4cd11f2dd.

Revert "Revert "Initial fix by adding field to FileInfo""

This reverts commit 6a810fd2cef1b1d8a7a1e96ac505c34e201acf1b.

WIP using resource UID and custom FileInfo field

Custom resource icons using dependencies approach

Add willnationsdev's feedback

- Refactored code to use ClassDB for inheritance checking.
- Removed redundant writing to fi.path

Attempt saving UID of script

This attempt failed because there's no way of fetching the UID of a script file. I assume this is because scripts aren't treated as a resource.

Revert back to script path approach

Hit a tough race condition

See editor_file_system.cpp:, line 1315, "TODO: Fix this workaround for a race condition."

Fix race condition & optimize filesystem_changed signal

Atlinx commented:
Ok, I've found a bug where the icons break if you move the script. It's likely due to resource_script_path not updating on change. I'll probably have to find some way to update all instances of a resource when their scripts change.

Atlinx commented:
It appears `.depren` is something that's supposed to exist as a temporary file. I cannot reproduce this issue on the current v3.4.3 version of Godot when I drag the scripts/resources around.

willnationsdev commented on editor/editor_file_system.cpp:
```diff
@@ -115,6 +115,17 @@ String EditorFileSystemDirectory::get_file_path(int p_idx) const {
+String EditorFileSystemDirectory::get_file_attached_script(int p_idx) const {
+	Vector<String> deps = get_file_deps(p_idx);
+	for (int i = 0; i < deps.size(); i++) {
+		if (deps[i].ends_with(".gd") || deps[i].ends_with(".cs")) {
```
Rather than checking for specific languages' extensions, it's better to use `ResourceLoader::get_resource_type(p_path)` to extract a typename and then use an inheritance check to see if it inherits from the `Script` class. If so, you know it's a script, regardless of what language it is.

You can *probably* assume that a user-defined type WON'T be an implementation of a Script, so you could skip it, but ordinarily, you would first check `ScriptServer::is_global_class()` and then based on that either use `EditorNode::get_editor_data().script_class_is_parent()` if true or `ClassDB::is_parent_class()` if false.

Atlinx replied:
I shouldn't have to worry about this running outside of the editor right? If so, then is there a way I can check if I'm in the editor?

willnationsdev replied:
@Atlinx Also, I just realized examining this, that this is not really a good solution (or at least, it doesn't seem very reliable). Any given file can have 0 or more dependencies, any number of which might happen to be a script. But you don't know *which* of those dependencies is the `script` property of the file in question. So, imo, you really *do* need to have some sort of resource script type property being tracked in the `FileInfo`/`FileCache` in `EditroFileSystem`. And rather than forcing the use of `ResourceFormatLoaderText`, you would need to define a virtual function in `ResourceFormatLoader` that gets the scripted type of the resource, if it has one, which you would then override in the various derived implementations. Then the logic to get the file's resource script type would involve iterating over all of the loaders present in the `ResourceLoader`.

Atlinx replied:
Does this mean we would have to break compatibility with 3.x? Adding another FileInfo means incrementing the file cache file version, and I'm not sure if this counts as compatibility breakage. Either way, I'll try to rewrite this PR using your suggestions.

willnationsdev replied:
I'm *guessing* it would involve breaking compatibility. Even if you assume that it's possible for 3.x and 4.x to have their own branching continuations of file cache versions, the fact that they share the same sequential numeric versioning scheme means that as soon as 3.x attempts to increment beyond where it already is, it will be stepping into the realm of 4.x's prior iterations (probably?). So yeah, I doubt this is something that can be properly done in 3.x, simply because the EditorFileSystem doesn't have any mechanism by which it can efficiently keep track of the type of the script that may or may not be assigned to a given resource file.

willnationsdev replied:
The ResourceFormatLoader thing might also become a headache *IF* you end up also modifying `get_resource_type` to return a possible non-C++ class as there are many calls to this method throughout the codebase, all of which likely expect it to return an in-engine C++ type rather than a potential scripted type. If you don't change the logic of that method though, and instead define a new function to access the extra information, thereby demanding that any scenario that need it specifically call that separate function, then there's likely to be less conflict with other uses.

willnationsdev commented on editor/filesystem_dock.cpp:
```diff
@@ -52,11 +52,21 @@
 	if (!p_is_valid) {
 		file_icon = get_theme_icon(SNAME("ImportFail"), SNAME("EditorIcons"));
 	} else {
+		if (p_file_type == "Resource") {
```
This assumes that all user-defined resources will be things that extend `Resource`. If I chose to implement a script extending, say, `Texture`, then I would be unable to make use of this. Nor if I chose to define a second type extending my first one. So,  rather than checking for an exact match, you have to perform another inheritance check, akin to the process I described in my earlier comment w/ ScriptServer, EditorData, and ClassDB.

willnationsdev replied:
@Atlinx Oh, also, when referring to the String values of type names, it's best to use `<type>::get_class_static()` rather than a hardcoded string. So, in this instance, you'd use `Resource::get_class_static()`.

willnationsdev commented on editor/filesystem_dock.cpp:
```diff
@@ -52,11 +52,21 @@
+			EditorFileSystemDirectory *script_dir = EditorFileSystem::get_singleton()->find_file(resource_script_path, &index);
+			if (script_dir != nullptr) {
+				file_icon = EditorNode::get_singleton()->get_class_icon(script_dir->get_file_script_class_name(index));
+				if (!file_icon.is_null())
+					return file_icon;
```
I believe Godot's conventions generally prefers curly braces, even when they aren't necessary (iirc).

willnationsdev commented on editor/filesystem_dock.cpp:
```diff
@@ -130,6 +140,7 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
 			FileInfo fi;
 			fi.name = p_dir->get_file(i);
 			fi.type = p_dir->get_file_type(i);
+			fi.path = p_dir->get_path().plus_file(fi.name);
```
From what I can tell, assigning this here is unnecessary(?). You only actually appear to use this value later on when attempting to get the filesystem path, but the way you use it - as `fi.path.get_base_dir()` - should be the same thing as `p_dir->get_path()` which is already assigned to the `lpath` variable declared earlier. And if you do end up needing the actual path, the `file_metadata` appears to already be doing that (and is set as the 0th slot metadata for the `file_item` `TreeItem`). Unless I'm misreading the code.

willnationsdev reviewed (changes requested):
I don't know yet whether the deps stuff is an adequate solution (cause I myself haven't researched that much), but I'll probably do some experimenting and get back to you if I discover any problems. I also still need to actually test that it works as expected, hehe. Otherwise, aside from the script and inheritance hardcoding, this is looking like a simple and elegant solution. Thanks!

willnationsdev commented:
@Atlinx Also...

> Does anyone know how Godot currently handles fixing dependencies when files are moved/deleted?

I'm not sure of this myself, but I know that the engine now has a UUID system for resources (the `ResourceUID::ID` type?). Perhaps you could leverage that to track the file? I'm not really sure yet whether that is safe to record in the filecache / EditorFileSystem though, so it would certainly warrant further investigation before determining whether that's the best path forward.

willnationsdev commented:
From what I can tell, the `ResourceUID` cache gets updated during `_update_scan_actions()` which processes all of the diffs between the last scan and the current scan. Then, after the filesystem has been re-generated and all files re-imported / UIDs re-cached, *that's* when `_queue_update_script_classes()` runs to trigger the re-assigning of all the various script class information to the file info. So - to me at least - it seems like it should be safe to use a ResourceUID::ID to refer to the script that the file is dependent on. That way, you'd be able to look up the path to the script regardless of whether it is renamed, moved, or modified such that the global class name changes.

Atlinx commented:
Do you have any ideas what the `.depren` files do? I'm still concerned that the original `.tres` not updating could break everything.

willnationsdev commented:
> Do you have any ideas what the .depren files do? I'm still concerned that the original .tres not updating could break everything.

I'm not quite sure. Usually things like that are used to quickly redirect any requests for a given file to wherever it was moved to. But I don't know yet why you would even need something like that if the ResourceUID stuff was added to help resolve issues with file paths moving around. The `.depren` stuff is new, so I'll need to investigate what the use case and methodology even is before I can give a definitive answer.

I'll just hold off on my own work until this PR is more refined I suppose since I'll be heavily relying on the ability to derive a resource's attached script's global class name from the EditorFileSystem in my Inspector optimizations for the exported resources PR I'm doing, (i.e. it'll build off of whatever it is you do).

Atlinx commented:
Yeah, I suspect `.depren` files were used as temp files in the past. For the scripts, do they even have a UID? Script's don't have a `tres` file, and looking in the `uid_cache.bin` file generated by Godot, I can only find `.tres` file paths.

willnationsdev commented:
> On another note, for the scripts, do they even have a UID?

Seems like that's the distinction. `ResourceFormatLoader` defines a virtual function, `get_resource_uid(p_path)`, and that method is overridden in both `ResourceFormatLoaderText` and `ResourceFormatLoaderBinary`, but there aren't any other overridden implementations that show up in a `Find All References` search. That said, it doesn't mean such resource files *can't* have `ResourceUID`s. It just means that none of them yet support it (as far as I can tell).

willnationsdev commented:
So basically, if `ResourceLoader::get_resource_uid` (which iterates over all the format loaders and calls their respective implementations) is able to derive a valid UID from the path, then you get to use it (yay!). Otherwise, you have to rely on something less effective.

Atlinx commented:
Actually it looks like UID are supported for scripts. A dig through `editor_file_system.cpp` revealed this bit of code under `_scan_new_dir`.
```C++
if (fi->uid != ResourceUID::INVALID_ID) {
	if (ResourceUID::get_singleton()->has_id(fi->uid)) {
		ResourceUID::get_singleton()->set_id(fi->uid, path);
	} else {
		ResourceUID::get_singleton()->add_id(fi->uid, path);
	}
}
```

Atlinx commented:
Ok, I've pushed a WIP branch with script UID fetching -- I'm going to go to sleep now and I'll revisit this PR again when I'm free. @willnationsdev Do you use Discord or the RocketChat by any chance? Having a direct line of communication might make it easier to talk things out and share our findings.

willnationsdev commented:
@Atlinx Yes, I'm on Discord regularly w/ the same username on the GodotEngine server. Just mention or DM me if you'd like.

willnationsdev commented on editor/editor_file_system.cpp:
```diff
@@ -115,6 +115,18 @@ String EditorFileSystemDirectory::get_file_path(int p_idx) const {
+	Vector<String> deps = get_file_deps(p_idx);
+	for (int i = 0; i < deps.size(); i++) {
+		String resource_type = ResourceLoader::get_resource_type(deps[i]);
+		if (ClassDB::is_parent_class(resource_type, "Script")) {
```
`Script::get_class_static()`

willnationsdev commented on editor/editor_file_system.cpp:
```diff
@@ -139,6 +151,34 @@ Vector<String> EditorFileSystemDirectory::get_file_deps(int p_idx) const {
+	String script_path_base_dir = script_path.get_base_dir();
+	String script_file = script_path.get_file();
+
+	if (script_path_base_dir == EditorFileSystemDirectory::get_path()) {
+		// If our resource script path happens to be in the same directory as our
+		// resource, then we can directly fetch the script's FileInfo.
+		int file_idx = EditorFileSystemDirectory::find_file_index(script_file);
+		ERR_FAIL_COND_V(file_idx < 0, "");
+		return files[file_idx]->script_class_name;
+	} else {
+		// If our resource script path is not in the same directory as our resource,
+		// we must fetch it's EditorFileSystemDirectory first, and then fetch the script's
+		// FileInfo from the newly fetched directory.
+		EditorFileSystemDirectory *dir = EditorFileSystem::get_singleton()->get_filesystem_path(script_path_base_dir);
+		int file_idx = dir->find_file_index(script_file);
+		ERR_FAIL_COND_V(file_idx < 0, "");
+		return dir->files[file_idx]->script_class_name;
+	}
```
1. (unless you were just adding temporary notes for yourself) Generally prefer not including comments unless it specifically is providing additional contextual details for how it interacts with other code elsewhere or perhaps explains why it chose a specific algorithm over another one. The code should be written in such a way that it is largely self-documenting. Overuse of comments leads to not-maintained "documentation" embedded in the code that quickly becomes stale as code is modified over time in large projects.
2. Because `get_path()` and `find_file_index` are member functions (i.e. methods) of the `EditorFileSystemDirectory` class, and we are inside of that class currently, you do not need to preface them with the class name scope; it should automatically call the local implementation. You only do this for `EditorFileSystem::get_singleton()` because that specifically fetches a static instance of the `EditorFileSystem` class and thus can only be accessed via a static getter function to begin with.

willnationsdev commented on editor/editor_file_system.cpp:
```diff
@@ -899,6 +942,9 @@ void EditorFileSystem::_scan_new_dir(EditorFileSystemDirectory *p_dir, Ref<DirAc
 					fi->type = "TextFile";
 				}
 				fi->uid = ResourceLoader::get_resource_uid(path);
+				if (fi->type == Resource::get_class_static()) {
```
`type` can be any C++ class, so it might be more suitable to use an inheritance check here, no? We want to record the script path even if the script is attached to something more derived than the base `Resource` type.

willnationsdev commented on scene/resources/resource_format_text.cpp:
```diff
@@ -907,6 +907,84 @@ Error ResourceLoaderText::rename_dependencies(Ref<FileAccess> p_f, const String
+// Fetches the script attached to a custom resource, reading as few lines as possible from the resource.
+// Please call ResourceLoaderText::open() before calling this.
+Error ResourceLoaderText::get_resource_script_path(String &r_script_path) {
+	if (res_type != "Resource") {
+		error_text = "Expected a user-defined resource file";
+		_printerr();
+		error = ERR_PARSE_ERROR;
+		return error;
+	}
```
Is the direct comparison to "Resource" correct/intentional? Like, does it seem like the `ResourceLoaderText` only applies to resources that extend the base `Resource` type, or could there be C++ types deriving `Resource` that still save as generic resource file types? I'm not sure. Just want to double-check.

Atlinx replied:
Just did some testing and it looks like the res_type is stored as whatever base type your script extends from. I've only extended resources in my use cases, so this never occurred to me.

willnationsdev commented on scene/resources/resource_format_text.cpp:
```diff
@@ -907,6 +907,84 @@ Error ResourceLoaderText::rename_dependencies(Ref<FileAccess> p_f, const String
 	return OK;
 }
 
+// Fetches the script attached to a custom resource, reading as few lines as possible from the resource.
```
This top comment is unnecessary.

willnationsdev commented on scene/resources/resource_format_text.cpp:
```diff
@@ -907,6 +907,84 @@ Error ResourceLoaderText::rename_dependencies(Ref<FileAccess> p_f, const String
+		return error;
+	}
+
+	while (true) {
+		if (next_tag.name != "ext_resource") {
+			break;
+		}
```
I may be misinterpreting this, but judging from the way this loop is designed, it seems to break out with the first `ext_resource` it finds. idk if that's actually guaranteed to refer to the `script` property of the resource though (is it?). Worst case scenario (and what I suspect), you'd have to jump to the serialized properties, find the one named `script`, identify which ID number of `ExtResource(<id>)` is assigned to it, and THEN work backward from there to find which `ext_resource` tag at the top has that ID.

willnationsdev reviewed (commented):
Thanks for all the quick work on this so far!

Atlinx committed:
Remove comments and update docs

Atlinx committed:
Remove comments and update docs

Atlinx committed:
Remove some comments & unecessary include

Atlinx committed:
Remove some comments & unecessary include

Atlinx commented:
With @willnationsdev's guidance, I've fixed up this PR. I've rebased my changes and it's ready for review. If this gets approved, we could potentially make a 3.x port, as even though I've increased `filesystem_cache`'s version, different editor versions will end up generating their own cache file if the existing one doesn't match, so it shouldn't matter too much.

Calinou commented on doc/classes/EditorFileSystem.xml:
```diff
@@ -86,6 +86,11 @@
 		</signal>
+		<signal name="script_classes_loaded">
+			<description>
+				Emitted once all scripts have been loaded. This happens after filesystem_changed is emitted.
```
```suggestion
				Emitted once all scripts have been loaded. This happens after [signal filesystem_changed] is emitted.
```

Calinou commented on doc/classes/ResourceFormatLoader.xml:
```diff
@@ -17,6 +17,13 @@
+			<return type="String" />
+			<argument index="0" name="path" type="String" />
+			<description>
+				If implemented, gets the path of a script attached to a resource. Scripts are only attached to a resource for custom classes that inherits from Resource.
```
Add a clickable reference:

```suggestion
				If implemented, gets the path of a script attached to a resource. Scripts are only attached to a resource for custom classes that inherits from [Resource].
```

Calinou commented on editor/editor_file_system.cpp:
```diff
@@ -1170,10 +1241,10 @@ void EditorFileSystem::scan_changes() {
 			sp.hi = 1;
 			sp.low = 0;
 			scan_total = 0;
+			// Accrues changess
```
Fix typo:

```suggestion
			// Accrues changes
```

Calinou commented on editor/editor_file_system.cpp:
```diff
@@ -1184,7 +1255,10 @@ void EditorFileSystem::scan_changes() {
+		// Accrues changes
 		thread_sources.start(_thread_func_sources, this, s);
+		// Because of multithreading, changes are applied in a check that occurs in the process loop.
+		// If the loop detects the multithread is done, then it applies the changes. See notification Process for more.
```
Use exact name to make searching easier:

```suggestion
		// If the loop detects the multithread is done, then it applies the changes. See NOTIFICATION_PROCESS for more.
```

Atlinx committed:
Update doc/classes/ResourceFormatLoader.xml

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update doc/classes/ResourceFormatLoader.xml

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update doc/classes/EditorFileSystem.xml

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update doc/classes/EditorFileSystem.xml

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

KoBeWi commented:
I get this crash when trying to open a project:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.alpha.custom_build (4ce4840fb3ee41aac3b8bb72ac6a194d9cd72ed7)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] CowData<EditorFileSystemDirectory::FileInfo *>::_get_size (C:\godot_source\core\templates\cowdata.h:82)
[1] CowData<EditorFileSystemDirectory::FileInfo *>::size (C:\godot_source\core\templates\cowdata.h:132)
[2] Vector<EditorFileSystemDirectory::FileInfo *>::size (C:\godot_source\core\templates\vector.h:90)
[3] EditorFileSystemDirectory::find_file_index (C:\godot_source\editor\editor_file_system.cpp:54)
[4] EditorFileSystemDirectory::get_file_resource_script_class_name (C:\godot_source\editor\editor_file_system.cpp:176)
[5] FileSystemDock::_get_tree_item_icon (C:\godot_source\editor\filesystem_dock.cpp:62)
[6] FileSystemDock::_create_tree (C:\godot_source\editor\filesystem_dock.cpp:155)
[7] FileSystemDock::_create_tree (C:\godot_source\editor\filesystem_dock.cpp:111)
[8] FileSystemDock::_create_tree (C:\godot_source\editor\filesystem_dock.cpp:111)
[9] FileSystemDock::_update_tree (C:\godot_source\editor\filesystem_dock.cpp:310)
[10] FileSystemDock::set_file_sort (C:\godot_source\editor\filesystem_dock.cpp:2984)
[11] EditorNode::_load_docks_from_config (C:\godot_source\editor\editor_node.cpp:4790)
[12] EditorNode::_load_docks (C:\godot_source\editor\editor_node.cpp:4636)
[13] EditorNode::_sources_changed (C:\godot_source\editor\editor_node.cpp:1044)
[14] call_with_variant_args_helper<EditorNode,bool,0> (C:\godot_source\core\variant\binder_common.h:241)
[15] call_with_variant_args<EditorNode,bool> (C:\godot_source\core\variant\binder_common.h:351)
[16] CallableCustomMethodPointer<EditorNode,bool>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[17] Callable::call (C:\godot_source\core\variant\callable.cpp:51)
[18] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1120)
[19] Object::emit_signal<bool> (C:\godot_source\core\object\object.h:773)
[20] EditorFileSystem::_notification (C:\godot_source\editor\editor_file_system.cpp:1331)
[21] EditorFileSystem::_notificationv (C:\godot_source\editor\editor_file_system.h:148)
[22] Object::notification (C:\godot_source\core\object\object.cpp:849)
[23] SceneTree::_notify_group_pause (C:\godot_source\scene\main\scene_tree.cpp:864)
[24] SceneTree::process (C:\godot_source\scene\main\scene_tree.cpp:455)
[25] Main::iteration (C:\godot_source\main\main.cpp:2739)
[26] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:777)
[27] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:175)
[28] _main (C:\godot_source\platform\windows\godot_windows.cpp:197)
[29] main (C:\godot_source\platform\windows\godot_windows.cpp:211)
[30] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:225)
[31] __scrt_common_main_seh (D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[32] BaseThreadInitThunk
-- END OF BACKTRACE --
```

Atlinx commented:
> (C:\godot_source\editor\editor_file_system.cpp:176

Can you send a me a minimal reproduction project?

KoBeWi commented:
[ReproductionProject.zip](https://github.com/godotengine/godot/files/8748557/ReproductionProject.zip)
It's a project I'm using for testing things, so it's full of random stuff. I didn't investigate what exactly is causing the crash, so it's not exactly minimal, but it's small.

Atlinx committed:
Exclude packed scenes from parsing

Atlinx committed:
Exclude packed scenes from parsing

Atlinx commented:
> [ReproductionProject.zip](https://github.com/godotengine/godot/files/8748557/ReproductionProject.zip) It's a project I'm using for testing things, so it's full of random stuff. I didn't investigate what exactly is causing the crash, so it's not exactly minimal, but it's small.

I found the issue: I have to account for PackedScenes not being extendable by the user scripts (else we would attempt to parse it for an attached script).

KoBeWi commented:
Ok it no longer crashes, but for some reason I don't see the icons:
![image](https://user-images.githubusercontent.com/2223172/169703728-e26347af-e8d1-497e-b295-8f5e7e680b16.png)
My class:
```GDScript
extends Resource
class_name Test, "res://icon.png"

@export var value = 1
```
![image](https://user-images.githubusercontent.com/2223172/169703799-66283c75-7291-4870-b326-4c2b2540b66d.png)

Atlinx commented:
Icons should be relative to the script file. If your using the reproduction
project, then icon.png is at the root level, where there aren't any scripts.

On Sun, May 22, 2022, 11:42 AM Tomek ***@***.***> wrote:

> Ok it no longer crashes, but for some reason I don't see the icons:
> [image: image]
> <https://user-images.githubusercontent.com/2223172/169703728-e26347af-e8d1-497e-b295-8f5e7e680b16.png>
> My class:
>
> extends Resourceclass_name Test, "res://icon.png"
>
> @export var value = 1
>
> [image: image]
> <https://user-images.githubusercontent.com/2223172/169703799-66283c75-7291-4870-b326-4c2b2540b66d.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/60606#issuecomment-1133922791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGBRPK77OMJPP2RV3FVNDTLVLJIW3ANCNFSM5UUHYNJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

KoBeWi commented:
Well, I did put the script in the root folder.
Also why only relative paths? Is this some technical limitation? The icon shows correctly everywhere else (see the second screenshot).

Atlinx commented:
Ah, I had thought you were opening up a pre-existing file but then I realized you're making a new one. I've reproduced this on my side so I'll look into it. 👍

Atlinx committed:
Fix missing not operator

Atlinx committed:
Fix missing not operator

Atlinx commented:
@KoBeWi I had dropped a not operator when I was moving all the inheritance checking into a helper method. Everything should work now.

KoBeWi commented on core/io/resource_loader.cpp:
```diff
@@ -773,6 +783,19 @@ ResourceUID::ID ResourceLoader::get_resource_uid(const String &p_path) {
+
+	for (int i = 0; i < loader_count; i++) {
+		String path = loader[i]->get_attached_script_path(local_path);
+		if (path != "") {
```
```suggestion
		if (!path.is_empty()) {
```
This is established way of checking if String is empty.

KoBeWi commented on doc/classes/ResourceFormatLoader.xml:
```diff
@@ -17,6 +17,13 @@
+			<return type="String" />
+			<argument index="0" name="path" type="String" />
+			<description>
+				If implemented, gets the path of a script attached to a resource. Scripts are only attached to a resource for custom classes that inherits from [Resource].
```
```suggestion
				If implemented, gets the path of a script attached to a resource. Scripts are only attached to a resource for custom classes that inherit from [Resource].
```
The second sentence sounds a bit confusing; could be reworded.

Atlinx replied:
Would this work?

If implemented, gets the path of a script attached to a resource. An attached script only exists for custom resources, which are user-defined classes that inherit from [Resource].

KoBeWi replied:
Yes, this is more clear.
Although not really true, as you can attach a script to any Resource via the inspector.

Atlinx replied:
Should we add a bit for that then?

KoBeWi replied:
I think something like
```suggestion
				If implemented, gets the path of a script attached to a resource. The script is usually attached when creating a new instance of a [Resource]-based custom class.
```
should be fine. Although that sentence could as well be removed, as it's not that important where the script comes from.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -139,6 +140,34 @@ Vector<String> EditorFileSystemDirectory::get_file_deps(int p_idx) const {
+	ERR_FAIL_INDEX_V(p_idx, files.size(), "");
+	String script_path = get_file_resource_script_path(p_idx);
+
+	if (script_path == "") {
```
```suggestion
	if (script_path.is_empty()) {
```

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -339,6 +369,7 @@ void EditorFileSystem::_save_filesystem_cache() {
 	_save_filesystem_cache(filesystem, f);
 }
 
+// Multithreaded helper method for scan()
```
```suggestion
// Multithreaded helper method for scan().
```

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1134,6 +1193,7 @@ void EditorFileSystem::_delete_internal_files(String p_file) {
 	}
 }
 
+// Multi-threaded threaded function for scan_changes()
```
```suggestion
// Multi-threaded threaded function for scan_changes().
```

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1170,10 +1230,10 @@ void EditorFileSystem::scan_changes() {
 			sp.hi = 1;
 			sp.low = 0;
 			scan_total = 0;
+			// Accrues changes
```
```suggestion
			// Accrues changes.
```

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1184,10 +1244,17 @@ void EditorFileSystem::scan_changes() {
 }
 
+bool EditorFileSystem::_is_script_extendable_resource(const StringName &p_class) {
+	return !ClassDB::is_parent_class(p_class, Script::get_class_static()) && !ClassDB::is_parent_class(p_class, PackedScene::get_class_static()) && ClassDB::is_parent_class(p_class, Resource::get_class_static());
```
`Resource` check should be at the beginning, as the other checks are useless if the classes aren't Resource.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1240,13 +1307,20 @@ void EditorFileSystem::_notification(int p_what) {
+					// Flush out the accrued ItemActions (if any) from _scan_filesystem(), and suppress any filesystem_changed signals.
+					suppress_filesystem_changed_signal = true;
 					_update_scan_actions();
-					emit_signal(SNAME("filesystem_changed"));
-					emit_signal(SNAME("sources_changed"), sources_changed.size() > 0);
+					suppress_filesystem_changed_signal = false;
+
 					_queue_update_script_classes();
+					_try_emit_filesystem_changed();
```
What's the reason for suppressing the signal if you are emitting it afterwards?

KoBeWi commented on editor/editor_file_system.h:
```diff
@@ -290,6 +301,8 @@ class EditorFileSystem : public Node {
 	bool _scan_extensions();
 	bool _scan_import_support(Vector<String> reimports);
 
+	bool _try_emit_filesystem_changed();
```
Why does this return `bool` if you never use the value?

Atlinx replied:
Most try methods I've seen follow the pattern of returning whether the attempt was successful or not. I ended up implementing that here.

KoBeWi replied:
If the value isn't used anywhere, it should be `void`. If you are that worried about conventions, you could just rename it.

Atlinx replied:
Alright, I'll just make it void return type.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -52,11 +52,19 @@
+		if (EditorFileSystem::_is_script_extendable_resource(p_file_type)) {
+			// Fetches the resourse icon using the EditorNode's script icon cache and EditorFileSystemDirectory's FileInfo cache.
+			String resource_script_name = p_dir->get_file_resource_script_class_name(p_file_index);
+			if (resource_script_name != "") {
```
```suggestion
			if (resource_script_name.is_empty()) {
```

KoBeWi commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1398,6 +1398,62 @@ ResourceUID::ID ResourceLoaderText::get_uid(Ref<FileAccess> p_f) {
 
+// Please call ResourceLoaderText::open() before calling this.
+// Returns empty string if the attached script could not be fetched.
+String ResourceLoaderText::get_attached_script_path(Ref<FileAccess> p_f) {
```
Fun fact: this method will return the icon of the first found Script-type sub-resource, not necessarily from the `script` property. However it's nearly impossible to get it to display wrong icon without editing the scene manually, so I wouldn't worry about it.

Atlinx replied:
Yea the intention was to exit early so we avoid parsing the whole resource file. I couldn't find any use case of having multiple Scripts, so that's why it's like that right now.

Atlinx replied:
But is there a way to check if a sub-resource maps to the `script` property?

Atlinx replied:
Ah, it looks like the script property just located on the bottom of the `[resource]`. I'll have to change the parsing to account for that.

KoBeWi replied:
As I said, this is not important in most cases. When you create a custom resource the script will always be first

KoBeWi replied:
I think it doesn't happen very often, but I don't use custom resources much, so maybe I'm wrong.

If you find a way to fix this without affecting performance and complicating the code, go ahead. If not, I'd just leave it and wait until someone opens an issue about it. There's a chance no one will run into this situation or if they do, they won't consider it a bug (it doesn't really affect anything).

Atlinx replied:
Could we use ClassDB::APIType to detect if a class is built-in?

willnationsdev replied:
@KoBeWi

> As I said, this is not important in most cases. When you create a custom resource the script will always be first

Did you perhaps mean that the script will always be *last*? Because I have a use case in my sample project for #48201 where an exported resource property itself uses a global script class that has another global script class property as its exported property, and in *that* use case, there are 2 `[ext_resource type="Script"...]` tags with the bottom one being the `script` property's `Script` (i.e. the "attached" script). Which then results in @Atlinx's code triggering a bug in my code.

To be specific, my demo has a `Tile2D` node that has a `terrain` property of type `Terrain`. It is possible to load a `WildTerrain` resource onto it (or an `UrbanTerrain`), and the `WildTerrain` has an exported array of `NaturalResource` resources. And in this use case, if I save e.g. a `wild_terrain_1.tres` file, that file needs both `NaturalResource` and `WildTerrain`, but it grabs the wrong one b/c `WildTerrain` is at the bottom and it stops after it finds the first one.

Either way, this PR needs to be amended to fix that.

KoBeWi commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1398,6 +1398,62 @@ ResourceUID::ID ResourceLoaderText::get_uid(Ref<FileAccess> p_f) {
+			String id = next_tag.fields["id"];
+
+			if (!path.contains("://") && path.is_relative_path()) {
+				// path is relative to file being loaded, so convert to a resource path
```
```suggestion
				// Path is relative to file being loaded, so convert to a resource path.
```

KoBeWi commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1398,6 +1398,62 @@ ResourceUID::ID ResourceLoaderText::get_uid(Ref<FileAccess> p_f) {
+			String path = next_tag.fields["path"];
+			String id = next_tag.fields["id"];
+
+			if (!path.contains("://") && path.is_relative_path()) {
```
```suggestion
			if (path.is_relative_path()) {
```
I think that part is already checked by `is_relative_path()`.

Atlinx replied:
I copied this bit from `load()` Should I remove that part from both `load()` and `get_attached_script_path()`?

KoBeWi replied:
I guess. The `is_relative_path()` checks for `:/`, which is to handle e.g. `C:/` on Windows, but this substring is also used in `res://`, so checking it twice is redundant.

KoBeWi commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1495,6 +1551,24 @@ ResourceUID::ID ResourceFormatLoaderText::get_resource_uid(const String &p_path)
+	String type = ResourceLoader::get_resource_type(p_path);
+
+	// We assume that only Resources can have attached scripts (excluding Scripts, which still extend Resource).
+	if (ClassDB::is_parent_class(type, Script::get_class_static()) || !ClassDB::is_parent_class(type, Resource::get_class_static())) {
```
Again, the condition can be swapped.

Atlinx replied:
Whoops I was supposed to use the helper method here

Atlinx committed:
Update core/io/resource_loader.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update core/io/resource_loader.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/filesystem_dock.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/filesystem_dock.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update scene/resources/resource_format_text.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update scene/resources/resource_format_text.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

KoBeWi commented on editor/editor_file_system.h:
```diff
@@ -236,7 +245,8 @@ class EditorFileSystem : public Node {
 	List<ItemAction> scan_actions;
 
-	bool _update_scan_actions();
+	void _update_scan_actions();
+	bool _update_scan_actions_helper();
```
Feels like these two method names should be swapped tbh, but that would add more changes, so it's not important.

Atlinx replied:
Yeah, a lot of the names in `editor_file_system` aren't too descriptive IMO. I have a previous commit that documented tons of code, but I removed most of the comments because @willnationsdev mentioned they may end up becoming junk comments if they aren't properly maintained alongside changes.

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Update editor/editor_file_system.cpp

Co-authored-by: Tomek <kobewi4e@gmail.com>

Atlinx committed:
Add rest of KoBeWi's suggestions

Atlinx committed:
Add rest of KoBeWi's suggestions

Atlinx committed:
Add more of KoBeWi's suggestions

- Use _is_script_extendable_resource helper in edior_file_system
- Change wording in docs
- Remove unecessary bool return value

Atlinx committed:
Add more of KoBeWi's suggestions

- Use _is_script_extendable_resource helper in edior_file_system
- Change wording in docs
- Remove unecessary bool return value

Atlinx commented:
@KoBeWi Can I move the helper `_is_extendable_script` method to ClassDB? It looks like keeping it any of the editor classes can cause the engine build to fail, due to the editor code not being included.

KoBeWi commented:
Ah yes, it should be moved. `scene` code shouldn't include editor stuff.

Atlinx committed:
Relocated helper method + add file list icon

Atlinx committed:
Relocated helper method + add file list icon

lyuma reviewed (commented):
Needs an implementation of `get_attached_script_path` in ResourceFormatLoaderBinary and ResourceLoaderBinary, for .res files

Neteraxe commented:
Any PR for core should care about performance and be really simple.
I have found you use i++ operator, could you make ++i not i++? 
In some compilers, ++i may is more efficient, and it correctly states your intent more than notes.

Atlinx commented:
> Actually, this PR is truly innovative and visual, and just a few small updates for the feature because any PR for core should care about performance and be really simple. For code form, I have found you use i++ operator, could you make ++i not i++? In some compilers, ++i may is more efficient, and it correctly states your intent more than notes.

I'm not sure if that's necessary since a small compiler-level optimization would have little impact on the overall cost of parsing a resource file, etc.

Atlinx committed:
WIP binary format

Atlinx committed:
WIP binary format

Neteraxe commented:
I just want to tell you that in many cases you should focus on performance for core contribution, because the core means the part code will be reused many times.
But the PR reviewer is other expert, l think you should work for their satisfaction.

KoBeWi commented:
Changing `i++` to `++i` has absolutely minimal impact on the performance. It's also not something we use in the codebase. If we were to change that, it should be done in another PR that would cover all areas, not a single class.

Neteraxe commented:
You could take my advice or not. It is not important, but it means something. It is just for code, not this PR contribution. And you could check your loop.
PR contributor should follow reviewer. I just want to mention to you that the core part is performance sensitive, in most cases. It is just advices and tips that you could adopt.

Neteraxe commented:
You could take my advice or not. It is not important, but it means something. It is just for code, not this PR contribution. And you could check your loop.
PR contributor should follow reviewer. I just want to mention to you that the core part is performance sensitive, in most cases. It is just advices, and tips that you could adopt.

Atlinx committed:
Add binary format

Atlinx committed:
Add binary format

Atlinx commented:
For the binary format implementation of `get_attached_script_path`, I followed the same strategy as the text format implementation, where I look for the first Script external resource.

Atlinx commented on core/io/resource.h:
```diff
@@ -39,11 +39,15 @@
-	static void register_custom_data_to_otdb() { ClassDB::add_resource_base_extension(m_ext, get_class_static()); } \
-	virtual String get_base_extension() const override { return m_ext; }                                            \
-                                                                                                                    \
+#define RES_BASE_EXTENSION(m_ext)                                        \
```
My visual studio is religiously following the clang format so it formats the existing code whenever I save `resource.h`

KoBeWi replied:
What is your clang-format version? IIRC there was some bug in the newest one (14 I think), so maybe this is the reason.

Atlinx replied:
I downloaded Clang from Visual

akien-mga replied:
You must have been running 14.0.0 because that's the version which does those unnecessary changes. They would need to be reverted.

Atlinx replied:
That's weird, even when I explicitly bind my clang format to the v12.0.0 exe, it still reformats the entire resource.h file. What version do you use?

Atlinx replied:
Also, I notice I'm getting the error "Unknown key: SpacesInLineCommentPrefix" whenever my clang-format runs.

willnationsdev reviewed (changes requested):
I have encountered a bug with this PR. Read below for details.

willnationsdev commented on scene/resources/resource_format_text.cpp:
```diff
@@ -1398,6 +1399,65 @@ ResourceUID::ID ResourceLoaderText::get_uid(Ref<FileAccess> p_f) {
+		String type = next_tag.fields["type"];
+
+		// Skip the tags until we find the "Script" tag
+		if (type == "Script") {
+			String path = next_tag.fields["path"];
+			String id = next_tag.fields["id"];
+
+			if (path.is_relative_path()) {
+				// Path is relative to file being loaded, so convert to a resource path.
+				path = ProjectSettings::get_singleton()->localize_path(local_path.get_base_dir().plus_file(path));
+			}
+
+			if (remaps.has(path)) {
+				path = remaps[path];
+			}
+
+			// Removed the parsed resources from the total
+			resources_total -= resource_current;
+			resource_current = 0;
+
+			return path;
+		}
+
+		error = VariantParser::parse_tag(&stream, lines, error_text, next_tag);
+
```
@KoBeWi

> As I said, this is not important in most cases. When you create a custom resource the script will always be first

Did you perhaps mean that the script will always be *last*? Because I have a use case in my sample project for #48201 where an exported resource property itself uses a global script class that has another global script class property as its exported property, and in *that* use case, there are 2 `[ext_resource type="Script"...]` tags with the bottom one being the `script` property's `Script` (i.e. the "attached" script). Which then results in @Atlinx's code triggering a bug in my code.

To be specific, my demo has a `Tile2D` node that has a `terrain` property of type `Terrain`. It is possible to load a `WildTerrain` resource onto it (or an `UrbanTerrain`), and the `WildTerrain` has an exported array of `NaturalResource` resources. And in this use case, if I save e.g. a `wild_terrain_1.tres` file, that file needs both `NaturalResource` and `WildTerrain`, but it grabs the wrong one b/c `WildTerrain` is at the bottom and it stops after it finds the first one.

Either way, this PR needs to be amended to fix that.

Atlinx commented on core/io/resource_format_binary.cpp:
```diff
@@ -865,6 +865,22 @@ String ResourceLoaderBinary::get_unicode_string() {
+	int main_resource_idx = internal_resources.size() - 1;
+	String path = internal_resources[main_resource_idx].path;
+
+	for (const ExtResource &E : external_resources) {
+		if (E.type == Script::get_class_static())
+			return E.path;
+	}
```
If the position isn't deterministic, we would have to parse the whole resource, as the only way to know what is the main script is by parsing the properties attached to the resource. For script resources, this appears at the end of the resource. It'll still be better than loading the entire resource, b/c that loads sub-resources as well.

Atlinx replied:
Also interestingly enough, nothing seems to protect the script property. That is, you can implement a script property in your property and Godot won't complain 👀

Atlinx committed:
Change resource parsing to explicitly look for script property

Atlinx committed:
Change resource parsing to explicitly look for script property

Atlinx commented:
@willnationsdev Bug should be fixed now via explicit parsing of script property. Tell me if it works for you!

Atlinx committed:
Merge branch 'godotengine:master' into feat/32706_custom_resource_icons

Atlinx committed:
Change Map to HashMap

Atlinx committed:
Remove old code

Atlinx committed:
Merge branch 'feat/32706_custom_resource_icons' of https://github.com/Atlinx/godot into feat/32706_custom_resource_icons

akien-mga commented on core/io/resource.cpp:
```diff
@@ -37,6 +37,7 @@
 #include "core/object/script_language.h"
 #include "core/os/os.h"
 #include "scene/main/node.h" //only so casting works
+#include "scene/resources/packed_scene.h"
```
`core` shouldn't depend on `scene`. The above one with `node.h` is tech debt that needs to be solved eventually.

willnationsdev replied:
I would guess that they would prefer to use a hard-coded string in this instance. The only other "safe" alternative would be to use some other sort of wrapper logic present in the scene context to handle the implementation of this function. Not sure if that would be preferred though since it ultimately makes the code more complicated - and there aren't any special cases beyond Script and PackedScene, so the effort of making it loosely coupled *and* statically typed probably isn't worth it. My thought anyway.

Atlinx replied:
Alright I've removed the dependency

Atlinx committed:
Remove dependency on scene and remove unecessary resource.h changes

willnationsdev commented:
@Atlinx It took a little while, but I managed to squash and rebase all of your commits on top of the latest master branch (there were a lot of conflicts within just this branch's own commit history, for some reason). Note that I made the following changes, in accordance with what existed previously:

- The `_try_emit_filesystem_changed` method still returns `void` rather than your modification for it to return `bool`, since the returned value is never actually used any time the method is called.
- Because other changes in the history have a method overload of `EditorFilesystemDock::_get_file_icon()` that takes a `String` rather than an `EditorFilesystemDirectory` + `int index`, I switched back your renaming of the method from `_get_file_icon` to `_get_tree_item_icon` so that the method overload would still work properly.
- I had to do an additional commit on top because some `Map<>` reference should've been `HashMap<>` (not sure how?).

If you would like to grab this version of the code, it is available in my `gdres-filesystem` branch which does *not* include any commits related to my exportable resources PRs.

As for whether it is fixed or not, my QuickLoad dialog now shows the `UrbanTerrain` resources, but not the `WildTerrain` resources. I think it is somehow related to the EditorFileSystem not re-scanning files properly as I've identified it not doing so when I expected it to, e.g. when the editor regains focus or when I made changes to a GDScript file's source code in the ScriptEditor. And if the scan doesn't run, then the "attached script" info won't update (my hypothesis anyway). I'm about to do a live stream of some more involved debugging.

Atlinx commented:
> As for whether it is fixed or not, my QuickLoad dialog now shows the `UrbanTerrain` resources, but not the `WildTerrain` resources. I think it is somehow related to the EditorFileSystem not re-scanning files properly as I've identified it not doing so when I expected it to, e.g. when the editor regains focus or when I made changes to a GDScript file's source code in the ScriptEditor. And if the scan doesn't run, then the "attached script" info won't update (my hypothesis anyway). I'm about to do a live stream of some more involved debugging.

Yea I was playing around in the editor and I noticed the file system doesn't rescan when a new resource is created. I wasn't sure if this was intentional or not (since rescaning might be slow) and so I didn't modify it.

willnationsdev commented:
@Atlinx Gotcha. Then we'll need to figure out whether that's intentional or not / what the decision-making there was.

Regardless, I found another bug. With your new implementation, any resource with a global script attached to it that has an exported property containing a user-defined Resource type ends up not successfully parsing tags during the `ResourceFormatLoaderText::_get_attached_script_path()` method. In my case, my `WildTerrain` type which extends `Terrain` and then `Resource`, has an exported property of type `NaturalResource` which extends `Resource`. The `NaturalResource` instance shows up as a `[sub_resource ...]` tag when saved as a `.tres`. It then triggers an error when you call `parse_tag` here:

```gdscript
	while (next_tag.name != "resource") {
		error = VariantParser::parse_tag(&stream, lines, error_text, next_tag);
		if (error) {
			return "";
		}
	}
```

...because it starts the parsing process on the `script` identifier *underneath* the actual tag. So it just returns an empty string and thus does not associate a script resource path with the resource file.

Atlinx commented:
So it's throwing the error because it's parsing something that's not a tag, since I assume what's below `[sub_resource]` are assign statements?

willnationsdev commented:
@Atlinx yes, that is correct.

Atlinx committed:
Add skip until tag parsing

Atlinx commented:
@willnationsdev I've pushed some changes -- let me know if the latest commit works for you.

Atlinx committed:
Merge branch 'master' into feat/32706_custom_resource_icons

Atlinx committed:
Fix missing description tag in doc

Atlinx committed:
Fix indentation to tabs in docs

Atlinx committed:
Merge branch 'godotengine:master' into feat/32706_custom_resource_icons

Atlinx committed:
Fix docs

Atlinx commented:
Sounds good 👍, I'll just keep this in limbo for now. 

I think the most elegant solution would be adding metadata to tres files, so important data can be accessed faster without parsing the whole file. Refactoring EditorNode and EditorFileSystem could also work, but then we'd have to deal with invalidation, and it still doesn't address the issue of loading the entire resource file in order to access a small portion of it.

willnationsdev reviewed (commented):
(Edit: lol, I started writing this before I saw that you had already responded and seem perfectly at ease with the feedback, but anyway, just in case, I hope you find this explanation helpful, haha)

I wanna thank you again for all the work you put into this PR. Even if it isn't accepted in its current state, it definitely gives a better starting point from which to create a more effective solution.

I've been in past situations where I did a bunch of work for a PR and received a similar response of my changes being too "invasive" or "hacky". At the time, I was quite disheartened and frustrated. This may or may not be the case for you, but if it is, I'm re-reviewing these changes through this lens to consider in what ways others may find fault with them. I hope my feedback can provide insight into what could be improved here and enrich the quality of your future work should you choose to continue making contributions.

- For things labeled as "comment bloat":

    Generally, Godot's source code only contains comments when it is meant to explain *why* the chosen code was selected vs. another approach since it usually is a contextual matter specific to Godot's codebase/behavior. That is, one needs to take into account other code in order to know why the current code is doing what it does.
    
    However, if the current code is self-explanatory or can be discerned simply by reading the code within the class/function and putting the pieces together yourself, then explanatory comments are avoided since they are a maintenance hazard; they can easily become outdated by future contributors' work, and they unnecessarily space out the code making it harder to read as much content within a single view.

- For things labeled as "uncertain/questionable core bloat*" or "core bloat?*":

    This is a pull request that relates to improving the behavior of things at the editor level. As such, the effect it should have on the engine's core should be kept to a minimum. However, in the case of this specific problem, it ultimately has to be resolved by extracting serialized data from resources which, in turn, must involve making changes to the way in which resources are serialized (a piece of core logic). So you can't really *avoid* making *some* core changes to implement a solution (I **think**). If there IS a way to do that, then we need to investigate that possibility and re-work the solution to support that. If not though, I don't personally see any issue with the way you've added the extra behavior to the `ResourceFormatLoader`(s). Rather, the more pertinent issue is probably...(next point)
    
- For things labeled as "core bloat**":

    Variant is an extremely core class since it basically touches every piece of the engine, so any changes made to it are done so with great care. Nonetheless, there are a variety of new methods you've added to its parser which I'm not entirely sure should be necessary just for the sake of implementing this feature.

    Furthermore, the way in which the type information is extracted using the tags is inherently "hacky" since it revolves around figuring out where the list of property tags are, iterating through all of the properties, and identifying which one corresponds to the object's "script" property. This in turn makes the logic entirely dependent on the structural makeup of the tag tree, as well as whether the object in question has a "script" property. Should a later alteration in the engine choose to modify how script information is associated with an object, the VariantParser, of all things, would now suddenly be broken (which would not be directly related to Object-handling code and thus not obvious).
    
    In addition, the algorithm introduces a linear-time operation over the number of properties the object serializes. This contrasts with how, in concept at least, the operation is not at all requesting aggregate data related to the object's properties. It would make more sense for it to be constant-time operation.

None of this is to suggest that you didn't have reasons for doing things the way you did. I honestly probably would've first thought to do something similar to what you did (which is probably why I didn't question it the first time I reviewed it). That is, don't make any major changes to the way data is tracked and just work with what is already there to accomplish what minimum changes are necessary to make the logic further down the pipeline work properly.

In theory, you are attempting to avoid making significant changes to the core. In practice though, given the lengths your implementation has to go through to get things to work as needed, a worrying number of changes are introduced.

In this PR, you introduce changes to the core variant logic *merely* to support this one higher-level feature. If left alone and accepted, such single-purpose solutions culminate in a codebase that has many small pieces in the core that are not actually used by or relevant to the core at all, thus leading to code bloat. It would be preferred if we could make the minimum number of changes possible to the core to expose whatever information is necessary for higher-level pieces of the codebase to implement the required logic. This shifts the bulk of the changes to later layers of the architecture and results in a simpler and more direct approach to achieving the same goal.

I appreciate you going through all the effort of writing and maintaining this branch. It will certainly be a big help in guiding the direction of the next attempt at fixing this issue. Great work!

Atlinx commented:
Superceded by https://github.com/godotengine/godot/pull/73611
